Diamond2 conversion

ptsOSL opened the pull request:
Changes to make the new DII lattice, 48.mat, work in the Virtac.

Relatively few changes were required.

ptsOSL committed:
Add octupole field, b3 to supported fields

ptsOSL committed:
Handle KickAngle for multipoles attached to correctors

ptsOSL committed:
Update THERING conversion script for DII lattice

This conversion script should work for both
diamond I and diamond II ringmodes. 48.mat is the
latest DII ringmode.

ptsOSL commented:
The CI failure is caused by a bug in pytac 0.6.2 which is being worked on

T-Nicholls commented on src/atip/sim_data_sources.py:
```diff
@@ -200,21 +202,23 @@ def _get_KickAngle(self, cell):
+            KickAngle needs to be returned from cell 0 of the applicable Polynom(A/B)
+            attribute and so a conversion must take place. For independent
+            Correctors KickAngle can be returned directly from the element's
+            KickAngle attribute without any conversion. This is becausess
```
extra s on because, also what's the deal with the indentation?

ptsOSL replied:
Removed the ss. It wasnt a full tab before, so i fixed that. Although this is a private member so sphinx doesnt render it for the docs anyway.

T-Nicholls commented on src/atip/sim_data_sources.py:
```diff
@@ -200,21 +202,23 @@ def _get_KickAngle(self, cell):
-        if isinstance(self._at_element, at.elements.Sextupole):
+        if isinstance(self._at_element, at.elements.Sextupole) or isinstance(
+            self._at_element, at.elements.Multipole
+        ):
```
`isinstance` can take a tuple of types so this could be simplified to: `if isinstance(self._at_element, (at.elements.Sextupole, at.elements.Multipole)):`

ptsOSL replied:
Changed this

T-Nicholls commented on src/atip/sim_data_sources.py:
```diff
@@ -239,7 +243,9 @@ def _set_KickAngle(self, cell, value):
-        if isinstance(self._at_element, at.elements.Sextupole):
+        if isinstance(self._at_element, at.elements.Sextupole) or isinstance(
+            self._at_element, at.elements.Multipole
+        ):
```
Likewise

ptsOSL replied:
Done

T-Nicholls reviewed (changes requested):
Looks fine to me, but obviously I need to see the corresponding changes to Pytac & Virtac

ptsOSL committed:
Refactored tests and added ringmode 48

Added a single place to define the ringmodes to tests.
Created a combined fixture for loading both a pytac and
atip lattice. Added a new test which tests just loading
of the atip lattice.

MJGaughran commented on src/atip/load_sim.py:
```diff
@@ -34,6 +34,7 @@ def load_from_filepath(
         name=pytac_lattice.name,
         energy=pytac_lattice.get_value("energy", units=pytac.PHYS),
     )
+    at_lattice.disable_6d()
```
Why is this now disabled? Was this line accidentally included with:
https://github.com/DiamondLightSource/atip/pull/54/commits/871eec4d625161a06ca2f2f8eaac4e46f672e7a7

ptsOSL replied:
Yes, that was accidental, ill remove it

ptsOSL replied:
Removed

MJGaughran commented on src/atip/sim_data_sources.py:
```diff
@@ -227,19 +229,19 @@ def _set_KickAngle(self, cell, value):
-           KickAngle attribute without any conversion. This is because
-           independent Correctors have a KickAngle attribute in our AT lattice,
-           but those attached to Sextupoles do not.
+        .. Note:: If the Corrector is attached to a Sextupole  or ocupole then
```
ocupole -> Octupole

You may also want to capitalise the 'octupole' in the other docstring

ptsOSL replied:
Fixed

T-Nicholls replied:
I think you capitalised the one above, but forgot this one

MJGaughran commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -12,8 +12,12 @@ function create_lattice_matfile(filename)
         if isempty(RING)
             global THERING;
             RING = THERING;
+            disp('Using THERING.');
```
Maybe explain why THERING is used? i.e. RING is empty

ptsOSL replied:
Updated

MJGaughran commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -51,17 +55,19 @@ function create_lattice_matfile(filename)
-            RING(y, :) = [];  % Delete hstrs that are preceded by a sextupole.
-        elseif (strcmp(RING{y, 1}.FamName, 'VSTR') && strcmp(RING{y-1, 1}.Class, 'Sextupole'))
-            RING(y, :) = [];  % Delete vstrs that are preceded by a sextupole.
+        if strcmp(RING{y, 1}.FamName, 'HSTR') && (strcmp(RING{y-1, 1}.Class, 'Sextupole') || strcmp(RING{y-1, 1}.Class, 'Multipole'))
```
You might want to check with Rick that this is also true for Multipole

ptsOSL replied:
This is true for all Octupoles in the O0X family

MJGaughran commented on tests/conftest.py:
```diff
@@ -62,25 +62,26 @@ def atlds():
-def at_lattice():
-    return atip.utils.load_at_lattice("HMBA")
+@pytest.fixture(scope="function", params=["I04"])
+def load_at_and_pytac_lattices(request):
```
All other fixtures here (and I believe in general) should be labelled as the noun it represents, rather than how you obtain it. I can see you've done the same thing with `get_lattice_filepath()`

ptsOSL replied:
Fixed this

ptsOSL committed:
Regenerated ring with new AT Markers

Rick added these

ptsOSL committed:
Remove mistakenly added disable_6d()

ptsOSL committed:
Change default lattice from RING to THERING

Also tidy up handling of missing RING and THERING

ptsOSL committed:
Remove if check for ThinCorrectorPass

This is not specified in any modern lattice files, so
we do not need to check for this anymore. Checked for 48, I04, DIAD, VMX

ptsOSL committed:
Check HSTR/VSTR has length 0 removing

Sanity check that a HSTR/VSTR after a sextupole has,
length 0, which should confirm it is part of the sextupole.

ptsOSL committed:
Rename test fixtures and add back pytac_lattice

T-Nicholls commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -1,70 +1,76 @@
 % If a filename is given that file will be updated to ATIP standard. Otherwise
-% the ring is taken from either of the 'RING' or 'THERING' global variables,
-% with 'RING' taking priority. If a filename is not passed the updated lattice
+% THERING is taken from either of the 'RING' or 'THERING' global variables,
+% with 'THERING' taking priority. If a filename is not passed the updated lattice
```
Remind me to have a chat with you about why the precedence was RING > THERING when I'm in the office tomorrow

ptsOSL replied:
okay, yeh I was curious, as by default mml defines THERING and not RING. I was thinking it was perhaps so that you dont overwrite THERING, but wasnt sure

ptsOSL replied:
Will add print statement explaining that we are updating THERING in place

T-Nicholls commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -1,70 +1,76 @@
-            RING(y, :) = [];  % Delete vstrs that are preceded by a sextupole.
+    while y < length(THERING)
+        % The data within the deleted elements is not needed
+        if strcmp(THERING{y, 1}.FamName, 'HSTR') && THERING{y, 1}.Length == 0 &&(strcmp(THERING{y-1, 1}.Class, 'Sextupole') || strcmp(THERING{y-1, 1}.Class, 'Multipole'))
```
Space between && and the bracket please

ptsOSL replied:
This is going to make the code so much more readable ;)

T-Nicholls commented on tests/conftest.py:
```diff
@@ -62,25 +62,31 @@ def atlds():
+def at_and_pytac_lattices(request):
+    lattices = []
+    lattices.append(load_csv.load(request.param, cs.ControlSystem()))
+    lattices.append(atip.utils.load_at_lattice(request.param))
```
This reminds me, I don't think we have any proper tests for the functions in utils, but they're pretty heavily integrated into our normal usage of ATIP and so should probably be tested

ptsOSL replied:
Yeah, there are also virtac tests to add sometime

ptsOSL committed:
Minor style fixes

ptsOSL committed:
Add warning message to mml conversion script

MJGaughran commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -1,70 +1,78 @@
     % Correct dimension order if necessary.
-    if size(RING, 1) == 1
-        RING = permute(RING, [2 1]);
+    if size(THERING, 1) == 1
```
I would recommend using a separate variable name here and in the rest of the file. If it is using RING from MML, it is afterwards referred to as THERING, which is confusing.

MJGaughran replied:
E.g. something like `mml_ring`

MJGaughran replied:
It also makes the comments confusing, since 'THERING' can sometimes refer to the MML THERING object, and at other times to our use of the variable in this script.

ptsOSL replied:
I have made some changes to this. We now use ATIP_RING. If a file containing an ATIP_RING is supplied, we use that, if not then we initialise ATIP_RING from THERING (if it exists). We then modify ATIP_RING and dont change THERING, finally we save ATIP_RING to either the file supplied or a new file called lattice.mat

MJGaughran commented on src/atip/rings/create_lattice_matfile.m:
```diff
@@ -1,70 +1,78 @@
+            disp('Using global THERING and modifying it in place.');
         end
+    else
+        disp('Using global THERING and modifying it in place.');
```
Why is this disp message a duplicate of the message above? Is there some way to simplify this logic?

ptsOSL replied:
Behaviour has changed and this is no longer the case

MJGaughran reviewed (commented):
The logic related to RING and THERING is confusing, and I'm not convinced we're unable to at least tidy it up a little bit.

Otherwise, you will want to get all tests to pass on this branch. I imagine the pytac PR will resolve this.

ptsOSL committed:
Update create_lattice to use ATIP_RING instead of RING

I also had to regenerate the three mat files so that they store
the variable ATIP_RING instead of RING

ptsOSL committed:
Minor changes for clearer code
